File: src/core/fountain-decoder.ts
import { PRNG, RobustSoliton } from './prng.js';
import { Frame } from './protocol.js';
import { MAX_FILE_SIZE } from './container.js';

interface DecoderNode {
  sequenceNumber: number;
  degree: number;
  neighbors: number[];      // Indices of source blocks this frame depends on
  payload: Buffer;          // The current XOR sum
}

/**
 * Luby Transform (LT) Fountain Decoder
 *
 * Reconstructs the original binary payload from a stream of fountain frames.
 * Uses a Belief Propagation / Ripple algorithm.
 */
export class FountainDecoder {
  private sessionId: number = -1;
  private blockCount: number = -1;
  private blockSize: number = -1;
  private totalPayloadLength: number = -1;
  private soliton: RobustSoliton | null = null;
  
  // State
  private frames: Map<number, DecoderNode> = new Map(); // Unsolved frames
  private decodedBlocks: Buffer[] = [];                 // Solved blocks
  private solvedCount: number = 0;                      // Number of solved blocks
  private isComplete: boolean = false;                  // True when all blocks are solved

  constructor() {}

  /**
   * Process a received frame.
   * Returns true if this frame caused the decoding to complete.
   */
  public receiveFrame(frame: Frame): boolean {
    if (this.isComplete) {
      return true;
    }

    const { header, payload } = frame;

    // Initialize session parameters on first valid frame
    if (this.sessionId === -1) {
      if (header.totalPayloadLength > MAX_FILE_SIZE) {
        throw new Error(`Payload length ${header.totalPayloadLength} exceeds maximum allowed (${MAX_FILE_SIZE} bytes)`);
      }
      if (header.blockCount * header.blockSize > MAX_FILE_SIZE + header.blockSize) {
        throw new Error('Block parameters exceed maximum allowed memory bounds');
      }

      this.sessionId = header.sessionId;
      this.blockCount = header.blockCount;
      this.blockSize = header.blockSize;
      this.totalPayloadLength = header.totalPayloadLength;
      this.decodedBlocks = new Array(this.blockCount).fill(null);
      this.soliton = new RobustSoliton(this.blockCount);
    } else {
      // Validate consistency
      if (
        this.sessionId !== header.sessionId ||
        this.blockCount !== header.blockCount ||
        this.blockSize !== header.blockSize ||
        this.totalPayloadLength !== header.totalPayloadLength
      ) {
        throw new Error('Inconsistent frame metadata received for current session');
      }
    }

    // Ignore duplicates
    if (this.frames.has(header.sequenceNumber)) {
      return false;
    }

    // Determine the neighbors (block indices) for this frame
    let d: number;
    let neighbors: number[];

    if (header.sequenceNumber < this.blockCount) {
      // Systematic frame: exactly matches block[sequenceNumber]
      d = 1;
      neighbors = [header.sequenceNumber];
    } else {
      // Repair frame: use PRNG and Soliton distribution
      const prng = new PRNG(header.sequenceNumber);
      d = this.soliton!.sampleDegree(prng);
      neighbors = this.selectDistinctIndices(prng, d, this.blockCount);
    }

    // Create the node
    let node: DecoderNode = {
      sequenceNumber: header.sequenceNumber,
      degree: d,
      neighbors: neighbors,
      payload: Buffer.from(payload), // copy
    };

    // Eliminate already solved blocks from this node's neighbors
    node = this.eliminateSolvedBlocks(node);

    // If it's already fully solved (degree 0 and we already had it), ignore
    if (node.degree === 0) {
      return false;
    }

    // Add to unsolved frames graph
    this.frames.set(header.sequenceNumber, node);

    // If this node is degree 1, we can solve a block and trigger the ripple effect
    if (node.degree === 1) {
      this.processRipple(node);
    }

    return this.isComplete;
  }

  /**
   * Remove already solved blocks from a new node.
   */
  private eliminateSolvedBlocks(node: DecoderNode): DecoderNode {
    const remainingNeighbors: number[] = [];
    for (const neighbor of node.neighbors) {
      if (this.decodedBlocks[neighbor]) {
        // Block is already known, XOR it out
        const block = this.decodedBlocks[neighbor];
        for (let i = 0; i < this.blockSize; i++) {
          node.payload[i] ^= block[i];
        }
      } else {
        remainingNeighbors.push(neighbor);
      }
    }
    node.neighbors = remainingNeighbors;
    node.degree = remainingNeighbors.length;
    return node;
  }

  /**
   * The belief propagation ripple effect.
   * When a block is solved, XOR it out of all other unsolved frames.
   */
  private processRipple(initialNode: DecoderNode) {
    const queue: DecoderNode[] = [initialNode];

    while (queue.length > 0) {
      const node = queue.shift()!;
      
      // Safety check: is it still degree 1?
      if (node.degree !== 1) continue;
      
      const blockIndex = node.neighbors[0];
      
      // If we somehow already solved this block, skip
      if (this.decodedBlocks[blockIndex]) {
        this.frames.delete(node.sequenceNumber);
        continue;
      }

      // 1. Solve the block
      this.decodedBlocks[blockIndex] = Buffer.from(node.payload);
      this.solvedCount++;
      
      // Remove from unsolved frames
      this.frames.delete(node.sequenceNumber);

      if (this.solvedCount === this.blockCount) {
        this.isComplete = true;
        return; // We're done!
      }

      // 2. Ripple to other frames
      // Find all frames that depend on this newly solved block
      for (const [seq, otherNode] of this.frames.entries()) {
        const neighborIdx = otherNode.neighbors.indexOf(blockIndex);
        if (neighborIdx !== -1) {
          // XOR the solved block out of the frame
          for (let i = 0; i < this.blockSize; i++) {
            otherNode.payload[i] ^= node.payload[i];
          }
          
          // Remove the neighbor
          otherNode.neighbors.splice(neighborIdx, 1);
          otherNode.degree--;

          // If the frame becomes degree 1, add it to the ripple queue
          if (otherNode.degree === 1) {
            queue.push(otherNode);
          } else if (otherNode.degree === 0) {
            // Frame is fully solved, but we already have all its blocks
            this.frames.delete(seq);
          }
        }
      }
    }
  }

  /**
   * Reconstruct the final original payload buffer.
   * Throws if decoding is not yet complete.
   */
  public reconstructPayload(): Buffer {
    if (!this.isComplete) {
      throw new Error(`Cannot reconstruct payload: missing ${this.blockCount - this.solvedCount} blocks`);
    }

    const fullBuffer = Buffer.concat(this.decodedBlocks);
    
    // Trim any padding from the last block using the original totalPayloadLength
    return fullBuffer.subarray(0, this.totalPayloadLength);
  }

  public getProgress(): number {
    if (this.blockCount === -1) return 0;
    return this.solvedCount / this.blockCount;
  }

  public getSolvedCount(): number {
    return this.solvedCount;
  }

  public getBlockCount(): number {
    return this.blockCount;
  }

  private selectDistinctIndices(prng: PRNG, d: number, max: number): number[] {
    const indices = new Set<number>();
    while (indices.size < d) {
      indices.add(prng.nextInt(0, max));
    }
    return Array.from(indices);
  }
}

File: mobile/src/DEQR.Core/Fountain/FountainDecoder.cs
using DEQR.Core.Container;
using DEQR.Core.PRNG;
using DEQR.Core.Protocol;

namespace DEQR.Core.Fountain;

internal sealed class DecoderNode
{
    public required uint SequenceNumber { get; init; }
    public required List<int> Neighbors { get; set; }
    public required byte[] Payload { get; init; }
    public int Degree => Neighbors.Count;
}

public sealed class FountainDecoder
{
    private long _sessionId = -1;
    private int _blockCount = -1;
    private int _blockSize = -1;
    private int _totalPayloadLength = -1;
    private RobustSoliton? _soliton;

    private readonly Dictionary<uint, DecoderNode> _frames = new();
    private readonly HashSet<uint> _seenSequences = new();
    private byte[]?[] _decodedBlocks = Array.Empty<byte[]?>();
    private int _solvedCount;
    private bool _isComplete;

    public bool IsComplete => _isComplete;
    public int SolvedCount => _solvedCount;
    public int BlockCount => _blockCount;

    public bool ReceiveFrame(ProtocolFrame frame)
    {
        ArgumentNullException.ThrowIfNull(frame);
        if (_isComplete) return true;

        FrameHeader header = frame.Header;
        byte[] payload = frame.Payload;
        ValidateFrameShape(header, payload);

        if (!_seenSequences.Add(header.SequenceNumber))
        {
            return false;
        }

        if (_sessionId == -1)
        {
            Initialize(header);
        }
        else if (_sessionId != header.SessionId ||
                 _blockCount != header.BlockCount ||
                 _blockSize != header.BlockSize ||
                 _totalPayloadLength != header.TotalPayloadLength)
        {
            throw new InvalidDataException("Inconsistent frame metadata received for current session");
        }

        List<int> neighbors;
        if (header.SequenceNumber < _blockCount)
        {
            neighbors = new List<int> { checked((int)header.SequenceNumber) };
        }
        else
        {
            var prng = new Mulberry32PRNG(header.SequenceNumber);
            int degree = _soliton!.SampleDegree(prng);
            neighbors = FountainEncoder.SelectDistinctIndices(prng, degree, _blockCount).ToList();
        }

        var node = new DecoderNode
        {
            SequenceNumber = header.SequenceNumber,
            Neighbors = neighbors,
            Payload = (byte[])payload.Clone()
        };

        EliminateSolvedBlocks(node);
        if (node.Degree == 0) return false;

        _frames[header.SequenceNumber] = node;
        if (node.Degree == 1) ProcessRipple(node);
        return _isComplete;
    }

    public byte[] ReconstructPayload()
    {
        if (!_isComplete)
        {
            throw new InvalidOperationException($"Cannot reconstruct payload: missing {_blockCount - _solvedCount} blocks");
        }

        byte[] fullBuffer = new byte[checked(_blockCount * _blockSize)];
        for (int i = 0; i < _blockCount; i++)
        {
            byte[] block = _decodedBlocks[i] ?? throw new InvalidOperationException($"Decoded block {i} is unexpectedly missing");
            Buffer.BlockCopy(block, 0, fullBuffer, checked(i * _blockSize), _blockSize);
        }

        byte[] result = new byte[_totalPayloadLength];
        Buffer.BlockCopy(fullBuffer, 0, result, 0, _totalPayloadLength);
        return result;
    }

    public double GetProgress() => _blockCount <= 0 ? 0 : (double)_solvedCount / _blockCount;

    private void Initialize(FrameHeader header)
    {
        _sessionId = header.SessionId;
        _blockCount = header.BlockCount;
        _blockSize = header.BlockSize;
        _totalPayloadLength = checked((int)header.TotalPayloadLength);
        _decodedBlocks = new byte[]?[_blockCount];
        _soliton = new RobustSoliton(_blockCount);
    }

    private static void ValidateFrameShape(FrameHeader header, byte[] payload)
    {
        if (header.BlockCount == 0) throw new InvalidDataException("Block count must be greater than zero");
        if (header.BlockSize == 0) throw new InvalidDataException("Block size must be greater than zero");
        if (header.TotalPayloadLength == 0) throw new InvalidDataException("Total payload length must be greater than zero");
        if (header.TotalPayloadLength > ContainerParser.MaxFileSize)
        {
            throw new InvalidDataException($"Payload length {header.TotalPayloadLength} exceeds maximum allowed ({ContainerParser.MaxFileSize} bytes)");
        }
        if (payload.Length != header.BlockSize)
        {
            throw new InvalidDataException($"Frame payload length {payload.Length} does not match declared block size {header.BlockSize}");
        }

        long paddedLength = (long)header.BlockCount * header.BlockSize;
        if (paddedLength > ContainerParser.MaxFileSize + header.BlockSize)
        {
            throw new InvalidDataException("Block parameters exceed maximum allowed memory bounds");
        }

        int expectedBlockCount = checked((int)((header.TotalPayloadLength + header.BlockSize - 1u) / header.BlockSize));
        if (expectedBlockCount != header.BlockCount)
        {
            throw new InvalidDataException($"Block count {header.BlockCount} is inconsistent with payload length {header.TotalPayloadLength} and block size {header.BlockSize}");
        }
    }

    private void EliminateSolvedBlocks(DecoderNode node)
    {
        var remaining = new List<int>(node.Neighbors.Count);
        foreach (int neighbor in node.Neighbors)
        {
            byte[]? solved = _decodedBlocks[neighbor];
            if (solved is null)
            {
                remaining.Add(neighbor);
                continue;
            }

            for (int i = 0; i < _blockSize; i++) node.Payload[i] ^= solved[i];
        }
        node.Neighbors = remaining;
    }

    private void ProcessRipple(DecoderNode initialNode)
    {
        var queue = new Queue<DecoderNode>();
        queue.Enqueue(initialNode);

        while (queue.Count > 0)
        {
            DecoderNode node = queue.Dequeue();
            if (node.Degree != 1) continue;

            int blockIndex = node.Neighbors[0];
            if (_decodedBlocks[blockIndex] is not null)
            {
                _frames.Remove(node.SequenceNumber);
                continue;
            }

            _decodedBlocks[blockIndex] = (byte[])node.Payload.Clone();
            _solvedCount++;
            _frames.Remove(node.SequenceNumber);

            if (_solvedCount == _blockCount)
            {
                _isComplete = true;
                return;
            }

            foreach ((uint sequence, DecoderNode otherNode) in _frames.ToList())
            {
                int neighborIndex = otherNode.Neighbors.IndexOf(blockIndex);
                if (neighborIndex < 0) continue;

                for (int i = 0; i < _blockSize; i++) otherNode.Payload[i] ^= node.Payload[i];
                otherNode.Neighbors.RemoveAt(neighborIndex);

                if (otherNode.Degree == 1) queue.Enqueue(otherNode);
                else if (otherNode.Degree == 0) _frames.Remove(sequence);
            }
        }
    }
}
